service/controller/clusterapi/v21/resource/machinedeploymentstatus/resource.go

```diff
@@ -65,11 +66,19 @@ func (r *Resource) ensure(ctx context.Context, obj interface{}) error {
 		return microerror.Mask(err)
 	}
 
+	var md *v1alpha1.MachineDeployment
+	{
+		md, err = r.cmaClient.ClusterV1alpha1().MachineDeployments(cr.Namespace).Get(cr.Name, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
 	{
 		r.logger.LogCtx(ctx, "level", "debug", "message", "checking if status of machine deployment needs to be updated")
 
-		replicasChanged := cr.Status.Replicas != int32(cc.Status.Worker.Nodes)
-		readyReplicasChanged := cr.Status.ReadyReplicas != int32(cc.Status.Worker.Ready)
+		replicasChanged := cr.Status.Replicas != int32(cc.Status.Worker[md.Labels[label.MachineDeployment]].Nodes)
+		readyReplicasChanged := cr.Status.ReadyReplicas != int32(cc.Status.Worker[md.Labels[label.MachineDeployment]].Ready)
```

Comment:
So far we did not have key functions for controller context methods. This can panic in theory but in practise to make the resource even work at all you need to run the `workercount` resource first. We have a lot of these implicit dependencies in our operators.

Reply:
It's weird. Why the type in controller context can't be right in the first place? Then we wouldn't have to assert it all over the place.

Reply:
I'm not blocking on that because it was like that already.

Reply:
Can you elaborate what the right type would be? I don't follow the problem here.

Reply:
I mean why `cr.Status.Replicas` & `cc.Status.Worker[md.Labels[label.MachineDeployment]].Nodes` aren't both `int32` or `int`?

Reply:
`cr` here is `MachineDeployment`. `v1alpha1` seems to be already gone so here is `v1alpha2` where it is `int32`: https://github.com/kubernetes-sigs/cluster-api/blob/master/api/v1alpha2/machinedeployment_types.go#L151-L154. I think we just used the types which are defined upstream to have an easier life here without conversion back and forth.

Reply:
Using upstream types makes sense to me. In that case `cc.Status.Worker[md.Labels[label.MachineDeployment]].Nodes` should be of type `int32` IMO.

Reply:
With you. Will change this.